sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIO.java

```diff
@@ -630,10 +630,11 @@ public void processElement(ProcessContext c) throws Exception {
       }
 
       @FinishBundle
-      public void finishBundle() throws Exception {
+      public void finishBundle(FinishBundleContext c) throws Exception {
         bigtableWriter.flush();
         checkForFailures();
         LOG.info("Wrote {} records", recordsWritten);
+        c.output(null, Instant.now(), GlobalWindow.INSTANCE);
```

Comment:
This will output a null for every bundle, not for every firing of a window of the collection being written. I recommend adding a test that meaningfully captures the expected behavior, ideally with usage of TestStream so you can check what happens if there's multiple fired panes. My hunch is that this will be far from trivial (definitely not just a few lines) but worth it :)

Reply:
I guess this falls under my comment above in the PR body.

The naive implementation I can see of this is to keep track in ProcessElement of every window seen, then output a record in FinishBundle for each.  Is that sufficient?  I don't have any real-world streaming workloads to test this on, which is why I haven't tried yet. :)

Reply:
Maybe do what you did in this PR, but on top of this PCollection, stick a Combine - e.g. Count.globally() or a Sample.any()? That will produce 1 value per firing. I don't know whether it will actually work, but it might - a unit test would show.

Reply:
Ok, I think I get what you're saying.
In my flows I have using this, I do exactly that, I do a `Combine.globally()` on the output from `BigtableIO.write` and then use that output to trigger my continuation logic.  However, I expect that'll only work when all writes in all windows have finished, since the output currently is in the global window.

If I did keep track of the windows written by a bundle and output them in FinishBundle, I expect other windowed aggregates would then work the same way.

I'll try to write up some unit tests around this, I think what we've laid out will work.

Reply:
Is your input in the global window? If yes, then so should be the output of BigtableIO.write; the global window, like all other windows, is allowed to fire multiple times, according to the specified trigger. If no, then why is the output in the global window? I don't think BTIO.write does any rewindowing.

Reply:
Yeah, I've only ever used this in batch mode, so everything is in the global window (by default at least, and I don't change it).